dbms/src/Functions/zCurveBase.h

```diff
@@ -0,0 +1,342 @@
+            ResultType res_bit = 1ull << (number_of_bits - 1);
+            for (ResultType bit = 1ull << (number_of_bits - (arity - argument_index)); bit; bit >>= arity)
+            {
+                if (z_value & bit)
+                {
+                    result |= res_bit;
+                }
+                res_bit >>= 1;
+            }
+            return result;
+        }
+        /// Get maximal and minimal possible bit representation of a given argument
+        /// when the values fall in a given range of z values.
+        /* For the i-th argument of a function with arity k in the range [L, R] we do roughly the following:
+         * Consider L and R as blocks of k bits, such that the bit corresponding to out argument
+         * is the last bit of every block. The first block (highest-order bits) can be smaller than k.
+         * We try to find the maximum value for the last bit of each block, going from the highest bits.
+         * If the current blocks are L_b and R_b we are basically choosing between L_b and L_b + 1, since
+         * it's useless to pick a higher value. Out of those two we want the one, that is odd
+         * (and thus ends with a set bit, which would increase the value of our argument),
+         * while bearing in mind that the value shouldn't be greater than R_b
+         * (this means only that we can't choose L_b + 1 when L_b = R_b)
+         * If the value L_x we choose is strictly (!) less than R_b, then we should set the remainder of R to ones,
+         * since we can put any bits in the next blocks, without R being a constraint.
+         * We proceed similarly when calculating the minimum possible value.
+         */
+        std::pair<ResultType, ResultType> getMinMaxPossibleBitValueOfArgument(
+                ResultType left,
+                ResultType right,
+                size_t argument_index,
+                size_t arity) const
+        {
+            auto left_min = left, left_max = left;
+            auto right_min = right, right_max = right;
+            static constexpr size_t number_of_bits = sizeof(ResultType) << 3;
+
+            ResultType max_first_block_value = (1ull << (arity - argument_index)) - 1;
+            ResultType get_first_block = max_first_block_value << (number_of_bits - arity + argument_index);
+            ResultType first_plus_one = 1ull << (number_of_bits - arity + argument_index);
+
+
+            ResultType left_block_min, left_block_max, right_block_min, right_block_max;
+
+            // Handling the first block for MAX
+            left_block_max = left_max & get_first_block;
+            right_block_max = right_max & get_first_block;
+            bool is_left_best = static_cast<bool>(left_block_max & first_plus_one);
+            if (left_block_max < right_block_max)
+            {
+                if (!is_left_best)
+                {
+                    left_max += first_plus_one;
+                    left_max &= ~(first_plus_one - 1);
```

Comment:
I think, 

if left_max < first_plus_one , then left_max == first_plus_one; if left_max >= first_plus_one then !is_left_best == False.

Is my understanding wrong？